solidity/contracts/hooks/CCIPHook.sol

```diff
@@ -0,0 +1,211 @@
+    /// @param _destinationChainSelector The selector of the destination chain.
+    modifier onlyAllowlistedDestinationChain(uint64 _destinationChainSelector) {
+        if (!allowlistedDestinationChains[_destinationChainSelector])
+            revert DestinationChainNotAllowlisted(_destinationChainSelector);
+        _;
+    }
+
+    // ============ Internal functions ============
+
+    /// @notice Construct a CCIP message.
+    /// @dev This function will create an EVM2AnyMessage struct with all the necessary information for sending a text.
+    /// @param _receiver The address of the receiver.
+    /// @param _callData The string data to be sent.
+    /// @param _feeTokenAddress The address of the token used for fees. Set address(0) for native gas.
+    /// @return Client.EVM2AnyMessage Returns an EVM2AnyMessage struct which contains information for sending a CCIP message.
+    function _buildCCIPMessage(
+        address _receiver,
+        bytes memory _callData,
+        address _feeTokenAddress
+    ) internal pure returns (Client.EVM2AnyMessage memory) {
+        // Create an EVM2AnyMessage struct in memory with necessary information for sending a cross-chain message
+        return
+            Client.EVM2AnyMessage({
+                receiver: abi.encode(_receiver), // ABI-encoded receiver address
+                data: abi.encode(_callData), // ABI-encoded payload
+                tokenAmounts: new Client.EVMTokenAmount[](0), // Empty array as no tokens are transferred
+                extraArgs: "",
+                // Set the feeToken to a feeTokenAddress, indicating specific asset will be used for fees
+                feeToken: _feeTokenAddress
+            });
+    }
+
+    /// @notice Sends data to receiver on the destination chain.
+    /// @notice Pay for fees in native gas.
+    /// @dev Assumes your contract has sufficient native gas tokens.
+    /// @param _destinationChainSelector The identifier (aka selector) for the destination blockchain.
+    /// @param _receiver The address of the recipient on the destination blockchain.
+    /// @param _callData The text to be sent.
+    /// @return messageId The ID of the CCIP message that was sent.
+    function _sendMessagePayNative(
+        uint64 _destinationChainSelector,
+        address _receiver,
+        bytes memory _callData
+    )
+        internal
+        onlyAllowlistedDestinationChain(_destinationChainSelector)
+        returns (bytes32 messageId)
+    {
+        // Create an EVM2AnyMessage struct in memory with necessary information for sending a cross-chain message
+        Client.EVM2AnyMessage memory evm2AnyMessage = _buildCCIPMessage(
+            _receiver,
+            _callData,
+            address(0)
+        );
+
+        // Get the fee required to send the CCIP message
+        uint256 fees = ccip_router.getFee(_destinationChainSelector, evm2AnyMessage);
+
+        if (fees > msg.value)
+            revert NotEnoughBalance(address(this).balance, fees);
```

Comment:
this should be deduped with quoteDispatch

Reply:
I didn't understand this comment.

Quote dispatch checks the fees required to send the CCIP message.

line 127 fetches the fees that will guarantee message success during the call.
Fees returned by `ccip_router` may differ during the call time, as they are not constant. 
Line 127-130 can be removed obviously, if this is the intended behaviour